EDIT: Ignore this message freely.

----

> The existing unittests (nested `test(T)`) don't work with 256-bit vectors, they are specialized for 128-bit and need to be generalized if you want to instantiate them with the additional ones.
> 

I don't know exactly to handle submodules. druntime is submodule in ldc and so I can't switch to my branch.
But that's not a real problem, I will reapply the commit there and then checkout it. And I built the druntime-test-runner-debug.

But, it seems that there is an error in the old unittests. Or maybe better, something doesn't works that works in DMD
```
core/simd.d(808): Error: invalid cast from `ubyte` to `void`
```
That points to this: https://github.com/ldc-developers/druntime/blob/4995c1150cd8f6fefc794e46f72ccfed65a6c13d/src/core/simd.d#L808 line.
And it points there even if I remove the new tests.
It seems that the problem is with `test!void16()` in `storeUnaligned` tests. If I remove this line, then there are weird linker problems. They're less if I add `-shared`. But still, there's one:
```
undefined reference to `_D2rt4util7utility10safeAssertFNbNiNfbMAyaMQemZv'
```
I guess it's in the `assert` in the unittests but I don't understand why.

As for the unittests per se, I surely see this problem:
https://github.com/ldc-developers/druntime/blob/4995c1150cd8f6fefc794e46f72ccfed65a6c13d/src/core/simd.d#L741
whici is it doesn't have enough space for 32 bytes vectors and it doesn't test all 32 alignments. I suppose when you say generalizing, you mean the array size and the alignments.

--- a/src/core/simd.d
+++ b/src/core/simd.d
@@ -775,17 +775,19 @@ version (DMD_LDC)
             test!ulong2();
             test!double2();
             test!float4();
+            /*
             test!void32();
             test!double4();
             test!float8();
             test!byte32();
             test!ubyte32();
             test!short16();
-            test!ushort16()
+            test!ushort16();
             test!int8();
             test!uint8();
             test!long4();
             test!ulong4();
+            */
         }
     }
 
@@ -824,7 +826,7 @@ version (DMD_LDC)
                 }
             }
 
-            test!void16();
+            //test!void16();
             test!byte16();
             test!ubyte16();
             test!short8();
@@ -835,17 +837,19 @@ version (DMD_LDC)
             test!ulong2();
             test!double2();
             test!float4();
+            /*
             test!void32();
             test!double4();
             test!float8();
             test!byte32();
             test!ubyte32();
             test!short16();
-            test!ushort16()
+            test!ushort16();
             test!int8();
             test!uint8();
             test!long4();
             test!ulong4();
+            */
         }
     }
 }

--- a/src/core/simd.d
+++ b/src/core/simd.d
@@ -664,41 +664,7 @@ version (D_SIMD)
 }
 else version (LDC)
 {
-    /**
-    Load unaligned vector from address.
-    Params:
-        p = pointer to vector
-    Returns:
-        Vector
-    */
-    pragma(inline, true)
-    V loadUnaligned(V)(const V* p)
-        if(is(V == __vector))
-    {
-        static import ldc.simd;
-        alias Element = typeof(V.array[0]);
-        return ldc.simd.loadUnaligned!V(cast(const(Element)*) p);
-    }
-
-    /**
-    Store vector to unaligned address.
-    This is a compiler intrinsic.
-    Params:
-       p = pointer to vector
-       value = value to store
-    Returns:
-       value
-    */
-    pragma(inline, true)
-    V storeUnaligned(V)(V* p, V value)
-        if(is(V == __vector))
-    {
-        static import ldc.simd;
-        alias Element = typeof(V.array[0]);
-        ldc.simd.storeUnaligned!V(value, cast(Element*) p);
-        return value;
-    }
-
+    public import ldc.simd : loadUnaligned, storeUnaligned;
     /**
     Emit prefetch instruction.
     Params:

--- a/src/core/simd.d
+++ b/src/core/simd.d
@@ -775,19 +775,6 @@ version (DMD_LDC)
             test!ulong2();
             test!double2();
             test!float4();
-            /*
-            test!void32();
-            test!double4();
-            test!float8();
-            test!byte32();
-            test!ubyte32();
-            test!short16();
-            test!ushort16();
-            test!int8();
-            test!uint8();
-            test!long4();
-            test!ulong4();
-            */
         }
     }
 
@@ -796,12 +783,12 @@ version (DMD_LDC)
     unittest
     {
         // Memory to store the vector to:
-        // Should have enough data to test all 16-byte alignments, and still
-        // have room for a 16-byte vector
-        ubyte[32] data;
+        // Should have enough data to test all 512-byte alignments, and still
+        // have room for a 512-byte vector
+        ubyte[1024] data;
 
-        // to test all alignments from 1 ~ 16
-        foreach (i; 0..16)
+        // to test all alignments from 1 ~ 512
+        foreach (i; 0..512)
         {
             ubyte* d = &data[i];
 
@@ -826,30 +813,48 @@ version (DMD_LDC)
                 }
             }
 
-            //test!void16();
-            test!byte16();
-            test!ubyte16();
-            test!short8();
-            test!ushort8();
-            test!int4();
-            test!uint4();
-            test!long2();
-            test!ulong2();
-            test!double2();
-            test!float4();
-            /*
-            test!void32();
-            test!double4();
-            test!float8();
-            test!byte32();
-            test!ubyte32();
-            test!short16();
-            test!ushort16();
-            test!int8();
-            test!uint8();
-            test!long4();
-            test!ulong4();
-            */
+            static string size1Tests(string T)()
+            {
+                string res;
+                res ~= "test!(__vector("~T~"[1]))();";
+                res ~= "test!(__vector("~T~"[2]))();";
+                res ~= "test!(__vector("~T~"[4]))();";
+                return res;
+            }
+
+            static string generalTests(string T, string size)()
+            {
+                string res;
+                res ~= "test!(__vector("~T~"[8 / "~size~"]))();";
+                res ~= "test!(__vector("~T~"[16 / "~size~"]))();";
+                res ~= "test!(__vector("~T~"[32 / "~size~"]))();";
+                res ~= "test!(__vector("~T~"[64 / "~size~"]))();";
+                res ~= "test!(__vector("~T~"[128 / "~size~"]))();";
+                res ~= "test!(__vector("~T~"[256 / "~size~"]))();";
+                res ~= "test!(__vector("~T~"[512 / "~size~"]))();";
+                return res;
+            }
+            mixin(size1Tests!("void")());
+            mixin(size1Tests!("byte")());
+            mixin(size1Tests!("ubyte")());
+
+            test!(__vector(short[2]))();
+            test!(__vector(ushort[2]))();
+
+            mixin(generalTests!("void", "1")());
+            mixin(generalTests!("byte", "1")());
+            mixin(generalTests!("ubyte", "1")());
+
+            mixin(generalTests!("short", "2")());
+            mixin(generalTests!("ushort", "2")());
+
+            mixin(generalTests!("int", "4")());
+            mixin(generalTests!("uint", "4")());
+            mixin(generalTests!("float", "4")());
+
+            mixin(generalTests!("long", "8")());
+            mixin(generalTests!("ulong", "8")());
+            mixin(generalTests!("double", "8")());
         }
     }
 }

--- a/src/ldc/simd.di
+++ b/src/ldc/simd.di
@@ -204,15 +204,24 @@ assert(v.array == a);
 template loadUnaligned(V)
 if(is(typeof(llvmVecType!V)))
 {
-    alias BaseType!V T;
-    enum llvmT = llvmType!T;
+    alias BaseType!V ElementType;
+
+    pragma(inline, true)
+    deprecated("This is the DMD interface, use it only if cross-compiling. Otherwise, please use the LDC interface.")
+    V loadUnaligned(const V* p)
+    {
+        // Forward to LDC version.
+        return loadUnaligned!V(cast(const(ElementType)*) p);
+    }
+    
+    enum llvmElementType = llvmType!ElementType;
     enum llvmV = llvmVecType!V;
     enum ir = `
-        %p = bitcast `~llvmT~`* %0 to `~llvmV~`*
+        %p = bitcast `~llvmElementType~`* %0 to `~llvmV~`*
         %r = load `~llvmV~`, `~llvmV~`* %p, align 1
         ret `~llvmV~` %r`;
 
-    alias __ir_pure!(ir, V, const(T)*) loadUnaligned;
+    alias __ir_pure!(ir, V, const(ElementType)*) loadUnaligned;
 }
 
 /**
@@ -228,13 +237,23 @@ assert(v.array == a);
 template storeUnaligned(V)
 if(is(typeof(llvmVecType!V)))
 {
-  alias BaseType!V T;
-  enum llvmT = llvmType!T;
-  enum llvmV = llvmVecType!V;
-  enum ir = `
-      %p = bitcast `~llvmT~`* %1 to `~llvmV~`*
-      store `~llvmV~` %0, `~llvmV~`* %p, align 1`;
-  alias __ir_pure!(ir, void, V, T*) storeUnaligned;
+    alias BaseType!V ElementType;
+
+    pragma(inline, true)
+    deprecated("This is the DMD interface, use it only if cross-compiling. Otherwise, please use the LDC interface.")
+    V storeUnaligned(V* p, V value)
+    {
+        // Forward to LDC version.
+        storeUnaligned!V(value, cast(ElementType*) p);
+        return value;
+    }
+
+    enum llvmElementType = llvmType!ElementType;
+    enum llvmV = llvmVecType!V;
+    enum ir = `
+        %p = bitcast `~llvmElementType~`* %1 to `~llvmV~`*
+        store `~llvmV~` %0, `~llvmV~`* %p, align 1`;
+    alias __ir_pure!(ir, void, V, ElementType*) storeUnaligned;
 }
 
 private enum Cond{ eq, ne, gt, ge }

